datafusion/src/logical_plan/plan.rs

```diff
@@ -354,6 +356,43 @@ impl LogicalPlan {
             | LogicalPlan::CreateExternalTable { .. } => vec![],
         }
     }
+
+    /// returns all `Using` join columns in a logical plan
+    pub fn using_columns(&self) -> Result<Vec<HashSet<Column>>, DataFusionError> {
+        struct UsingJoinColumnVisitor {
+            using_columns: Vec<HashSet<Column>>,
+        }
+
+        impl PlanVisitor for UsingJoinColumnVisitor {
+            type Error = DataFusionError;
+
+            fn pre_visit(&mut self, plan: &LogicalPlan) -> Result<bool, Self::Error> {
+                if let LogicalPlan::Join {
+                    join_constraint: JoinConstraint::Using,
+                    on,
+                    ..
+                } = plan
+                {
+                    self.using_columns.push(
+                        on.iter()
+                            .map(|entry| {
+                                std::iter::once(entry.0.clone())
+                                    .chain(std::iter::once(entry.1.clone()))
```

Comment:
why not just flatmap with a `vec![entry.0.clone(), entry.1.clone()]` - it's cleaner

Reply:
i wanted to avoid an extra memory allocation incurred by `Vec::new`, but i will double check to see if once chain is actually generating the optimal code without memory allocations.

Reply:
@Jimexist sent https://github.com/apache/arrow-datafusion/pull/704 to address this. I found a comprise that's more readable and avoids the extra memory allocation.

Here is my test to compare the code gen with different approaches: https://godbolt.org/z/j3dcjbnvM.